hbase-server/src/main/java/org/apache/hadoop/hbase/compactionserver/CSRpcServices.java

```diff
@@ -0,0 +1,186 @@
+    try {
+      // use final bindAddress for this server.
+      return RpcServerFactory.createRpcServer(server, name, getServices(), bindAddress, conf,
+        rpcSchedulerFactory.create(conf, this, server), reservoirEnabled);
+    } catch (BindException be) {
+      throw new IOException(be.getMessage() + ". To switch ports use the '"
+          + HConstants.COMPACTION_SERVER_PORT + "' configuration property.",
+          be.getCause() != null ? be.getCause() : be);
+    }
+  }
+
+  void start() {
+    rpcServer.start();
+  }
+
+  protected Class<?> getRpcSchedulerFactoryClass() {
+    final Configuration conf = compactionServer.getConfiguration();
+    return conf.getClass(COMPACTION_SERVER_RPC_SCHEDULER_FACTORY_CLASS,
+      SimpleRpcSchedulerFactory.class);
+  }
+
+  public CSRpcServices(final HCompactionServer cs) throws IOException {
+    final Configuration conf = cs.getConfiguration();
+    compactionServer = cs;
+
+    final RpcSchedulerFactory rpcSchedulerFactory;
+    try {
+      rpcSchedulerFactory = getRpcSchedulerFactoryClass().asSubclass(RpcSchedulerFactory.class)
+          .getDeclaredConstructor().newInstance();
+    } catch (NoSuchMethodException | InvocationTargetException | InstantiationException
+        | IllegalAccessException e) {
+      throw new IllegalArgumentException(e);
+    }
+    // Server to handle client requests.
+    final InetSocketAddress initialIsa;
+    final InetSocketAddress bindAddress;
+
+    String hostname = getHostname();
+    int port =
+        conf.getInt(HConstants.COMPACTION_SERVER_PORT, HConstants.DEFAULT_COMPACTION_SERVER_PORT);
+    // Creation of a HSA will force a resolve.
+    initialIsa = new InetSocketAddress(hostname, port);
+    bindAddress =
+        new InetSocketAddress(conf.get("hbase.compaction.server.ipc.address", hostname), port);
+    if (initialIsa.getAddress() == null) {
+      throw new IllegalArgumentException("Failed resolve of " + initialIsa);
+    }
+    // Using Address means we don't get the IP too. Shorten it more even to just the host name
+    // w/o the domain.
+    final String name = cs.getProcessName() + "/"
+        + Address.fromParts(initialIsa.getHostName(), initialIsa.getPort()).toStringWithoutDomain();
+    // Set how many times to retry talking to another server over Connection.
+    ConnectionUtils.setServerSideHConnectionRetriesConfig(conf, name, LOG);
+    rpcServer = createRpcServer(cs, rpcSchedulerFactory, bindAddress, name);
+    final InetSocketAddress address = rpcServer.getListenerAddress();
+    if (address == null) {
+      throw new IOException("Listener channel is closed");
+    }
+    // Set our address, however we need the final port that was given to rpcServer
+    isa = new InetSocketAddress(initialIsa.getHostName(), address.getPort());
+    rpcServer.setErrorHandler(this);
+  }
+
+  @Override
+  public boolean checkOOME(final Throwable e) {
```

Comment:
I believe we have this method at ohter places too, is it possible to reuse the code instead of copy the code?

Reply:
Yes, Let me try to reduce the code